walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +40,263 @@ public int compare(Event a, Event b) {
+     */
+    Collection<TimeRange> resolveBestTime() {
+      ArrayList<TimeRange> mandatoryAttendeesMeetingTimes =
+          getMandatoryAttendeesMeetingTimes(
+              request.getDuration(), new HashSet<String>(request.getAttendees()));
+      Collection<TimeRange> optimalMeetingTimes =
+          getOptimalMeetingTimes(
+              mandatoryAttendeesMeetingTimes,
+              getChangesInOptionalAttendeesAttendance(
+                  getOptionalAttendeesFreeTimes(
+                      new HashSet<String>(request.getOptionalAttendees()), request.getDuration())),
+              request.getDuration());
+
+      // If there are no meeting times with at least one optional attendee, just return
+      // mandatoryAttendeesMeetingTimes.
+      return optimalMeetingTimes.isEmpty() ? mandatoryAttendeesMeetingTimes : optimalMeetingTimes;
     }
-    Collections.sort(events, ORDER_BY_START_ASC);
-
-    // Add first gap.
-    addIfLongEnough(
-        TimeRange.fromStartEnd(0, events.get(0).getWhen().start(), false),
-        possibleMeetingTimes,
-        request.getDuration());
-    int end = events.get(0).getWhen().end();
-    for (Event event : events) {
-      // event can be merged with current time range
-      if (event.getWhen().start() <= end) {
-        end = Math.max(end, event.getWhen().end());
-        continue;
+
+    /**
+     * Returns all meeting times that allow the most optional attendees to attend. These times must
+     * also fall in a time satisfying all mandatory attendees and must be at least minTime long.
+     * Utilizes two pointers: one to mandatoryAttendeesMeetingTimes, one to changeLog.
+     *
+     * @param mandatoryAttendeesMeetingTimes all meeting times satisfying mandatory attendees
+     * @param changeLog a change log of the number of available optional attendees over time
+     * @param minTime the minimum time a meeting must last for
+     */
+    private Collection<TimeRange> getOptimalMeetingTimes(
+        ArrayList<TimeRange> mandatoryAttendeesMeetingTimes,
+        TreeMap<Integer, Integer> changes,
+        long minTime) {
+
+      ArrayList<TimeRange> optimalMeetingTimes = new ArrayList<TimeRange>();
+      ArrayList<Map.Entry<Integer, Integer>> changeLog =
+          new ArrayList<Map.Entry<Integer, Integer>>(changes.entrySet());
+
+      // changeLogIndexis a pointer in changeLog,  mandatoryAttendeesMeetingTimesIndex is a pointer
+      // in mandatoryAttendeesMeetingTimes.
+      int mandatoryAttendeesMeetingTimesIndex = 0,
+          currAttendance = 0,
+          bestAttendance = 0,
+          prevTime = 0;
+      for (int changeLogIndex = 0; changeLogIndex < changeLog.size(); ++changeLogIndex) {
+        // Need to back up  mandatoryAttendeesMeetingTimesIndex in case we missed a time range in
+        // mandatoryAttendeesMeetingTimes.
+        mandatoryAttendeesMeetingTimesIndex = Math.max(0, mandatoryAttendeesMeetingTimesIndex - 1);
```

Comment:
This is not the clearest way to do this.

From what I can tell, you're intending to iterate over mandatoryAttendeesMeetingTimes only once, and you iterate over it over the course of iterating through changes. This "backtracking" is necessary because of the way you're iterating on the index variable.

The pattern I'd use for this is to (1) retain an iterator for mandatoryAttendeesMeetingTimes and (2) retain the "next" meeting time to be processed. Upon successful processing, get the `next` meeting time from the iterator, until there are no more remaining.

A skeleton of this:

```java
Iterator<TimeRange> mandatoryMeetingTimesIter = mandatoryAttendeesMeetingTimes.iterator();
Optional<TimeRange> nextMandatoryMeetingTimeOpt = 
    mandatoryMeetingTimesIter.hasNext() ? Optional.of(mandatoryMeetingTimesIter.next()) : Optional.empty();
for (Map.Entry<String,String> changeEntry : changes.entrySet()) {
  while (nextMandatoryMeetingTimeOpt.isPresent()) {
    TimeRange thisMandatoryMeetingTime = nextMandatoryMeetingTImeOpt.get();
    // TODO: Explain what this check is doing
    if (thisMandatoryMeetingTime.start() <  changeEntry.getKey()) {
      break;
    }
    updateOptimalTimes(
        TimeRange.fromStartEnd(
            Math.max(thisMandatoryMeetingTime.start(),  prevTime),
            Math.min(thisMandatoryMeetingTime.end(),
                     (Integer) changeLog.get(changeLogIndex).getKey()),
                     false));
    // Queue up the next meeting time.
    nextMandatoryMeetingTimeOpt = mandatoryMeetingTimesIter.hasNext() ?
                                                          Optional.of(mandatoryMeetingTimesIter.next()) : Optional.empty();
  }
  prevTime = (Integer) changeEntry.getKey();
  currAttendance += (Integer) changeEntry.getValue();
}
// TODO: What does it mean if nextMandatoryMeetingTimeOpt.isPresent()? Handle this
```

Reply:
1. l changed `thisMandatoryMeetingTime.start() <  changeEntry.getKey()` to `thisMandatoryMeetingTime.start() >=  changeEntry.getKey()` for logic reasons.
2. l don't see how `mandatoryMeetingTimesIter` is going back by 1 at the start of the outer for loop. It also fails some tests, because of this reason (ex. onlyOptionalAndDisjoint test). 

l could accomplish this decrement by keeping a `prevMandatoryMeetingTime` and setting `thisMandatoryMeetingTime` to that at the top of every outer for loop but that seems clumsy. Or l could do something with a ListIterator, since it supports a previous() method. Any thoughts?

Reply:
I don't really follow your comment, particularly this

> l don't see how mandatoryMeetingTimesIter is going back by 1 at the start of the outer for loop. It also fails some tests, because of this reason (easiest to see is in the onlyOptionalAndDisjoint test).

This could be interpreted as "why doesn't your code do this" or "my code doesn't actually do this". I'll try to answer both, but I just wanted to give you this caveat that i didn't really actually understand your comment.

The solution I described doesn't require mandatoryMeetingTimesIter to "go back one", because it is not being incremented unless the value is consumed. This is the purpose behind storing "the next value", and incrementing that immediately after the call to updateOptimalTimes. This more closely matches the semantics you want, rather than the for loop construction, which always increments the value when the value (that it is an index to) doesn't always get consumed.
gg

Reply:
Sorry for the confusion. 
1. The change in my first point was done locally and l didn't push that because it was failing tests.
2. The crux is that l need to advance mandatoryMeetingTimesIter enough for it to be too far in the future but then still be able to go back to last one that was valid (that time could still be valid for the next one in the changeLog). l think l get what you're saying now, with looking ahead to see the next mandatoryMeetingTime and choosing to advance or not. But in your code, when you "Queue up the next meeting time", you advance the iterator and you're unable to go back = bugs. l can use your iterator method but l'll have to make some changes from the code you proposed.

Reply:
Nvm, iterators are annoyingly verbose. l'll take you up on the `while` loop idea.